app/controllers/sessions_controller.rb

```diff
@@ -0,0 +1,93 @@
+class SessionsController < ApplicationController
+  def create
+    user = User.from_omniauth(request.env['omniauth.auth'])
+    session[:user_id] = user.id
+
+    redirect_to(
+      admin_path,
+      notice: t('omniauth_callbacks.success')
+    )
+  end
+
+  def destroy
+    if params[:SAMLRequest]
+      idp_logout_request
+    elsif params[:SAMLResponse]
+      validate_slo_response
+    else
+      sp_logout_request
+    end
+  end
+
+  def setup
+    if params.key?(:loa)
+      request.env['omniauth.strategy'].options[:authn_context] = [
+        "http://idmanagement.gov/ns/assurance/loa/#{params[:loa]}",
+        'http://idmanagement.gov/ns/requested_attributes?ReqAttr=email,first_name,last_name'
+      ]
+    end
+    render text: 'Omniauth setup phase.', status: 404
+  end
+
+  private
+
+  def idp_logout_request
+    logout_request = OneLogin::RubySaml::SloLogoutrequest.new(
+      params[:SAMLRequest],
+      settings: saml_settings
+    )
+    if logout_request.is_valid?
+      redirect_to_logout(logout_request)
+    else
+      render_logout_error
+    end
+  end
+
+  def saml_settings
+    @_saml_settings ||= OneLogin::RubySaml::Settings.new(SAML_SETTINGS)
+  end
+
+  def redirect_to_logout(logout_request)
+    Rails.logger.info "IdP initiated Logout for #{logout_request.nameid}"
+    sign_out
+    logout_response = OneLogin::RubySaml::SloLogoutresponse.new.create(
+      saml_settings,
+      logout_request.id,
+      nil,
+      RelayState: params[:RelayState]
+    )
+    redirect_to logout_response
+  end
+
+  def render_logout_error(logout_request)
+    error_msg = "IdP initiated LogoutRequest was not valid: #{logout_request.errors}"
+    Rails.logger.error error_msg
+    render inline: error_msg
+  end
+
+  def validate_slo_response
```

Comment:
this was also very confusing to me at first! while I am generally allergic to acronyms, this one seems pretty broadly accepted in the "single log out" world. do you think we should use full name, @adelevie ?